### ISSUE TITLE
Control over Learning Rate and/or initial step size in linesearch

### ISSUE THREAD

**Issue by matillda123** [patrick-kidger/optimistix#108]
Hi, 

is there a convenient way to influence the step size of the available minimizers/least_square solvers? (E.g. in optax one has the learning_rate argument for all optimizers)
So far with the optimistix solvers I was only able to gain some control by changing `solver.search.__dict__["step_init"]`, which doesnt seem right.

I guess one could just define a custom solver. But it seems weird that on one side there is so much customizability with the modular approach of optimistix, while with the predefined solvers there is very little user-control.

**Comment by patrick-kidger** [comment]
@matillda123 do you have an example of the kind of thing you'd like to do, using a particular solver for clarity? :)

**Comment by matillda123** [comment]
I'm not sure if it's possible to give a short example here. What im trying to do is basically fitting of a high-dimensional function to some data. The function basically consists of some FFTs.
Its similar to what is described in this [paper](https://opg.optica.org/oe/fulltext.cfm?uri=oe-27-13-18653&id=414640).

I'm not using one specific solver, rather I'm trying to get a feeling which solvers work nicely for this problem
I have the suspicion, that some solvers may sometimes work "significantly" better when the step size is biased towards bigger or smaller stepsizes. I noticed something like that in optax when tuning `learning_rate` with adam(+linesearch) and lbfgs.
So, I was looking for a way to test this with the predefined optimistix solvers and didn't find one.

I hope its clearer now what the issue is.

**Comment by johannahaffner** [comment]
I see, tough optimisation problem! 

It sounds like you want to do hyperparameter tuning - what do you get if a solver is not working well? Convergence to a local minimum, running out of steps, divergence?

And are you working with complex numbers, or have you split the problem into complex and real parts?

**Comment by matillda123** [comment]
I have real inputs but I use those to make complex numbers.
If a solver doesn't work well, for me, that generally means it's stuck in a local minimum or converging really slowly.

**Comment by johannahaffner** [comment]
So optimistix only ever sees real numbers? 

The functions you are working with can have several local minima, right? In this case, you might want to try multi-starting your optimisation. 
If you'd like step sizes larger than one wherever your optimisation landscape looks suitably quadratic, maybe try `optx.BFGS` with a `ClassicalTrustRegion`? The trust region radius can get larger than one, and has the widest range of step sizes, since these are updated multiplicatively. You can also take a look at the combinations here, such as

https://github.com/patrick-kidger/optimistix/blob/dcafc480d4a2bb9ec3b62658d5ae136db847b2d8/tests/helpers.py#L126

(It sounds like you tried `BacktrackingArmijo`, which will start with the full Newton step in BFGS and then chop that in half until it's satisfied.) 

Those are but two suggestions - I'm afraid I can't say much more with only a basic idea of what your problem is.

**Comment by matillda123** [comment]
Yes, optimistix should only see real numbers.

So far I only used standard solvers (mainly BFGS, NonlinearCG, GaussNewton and Dogleg). I think Dogleg has a trustregion linesearch. But it's tricky as that is one of the cases which can get stuck in local minima. 

Alright, so I guess there is no way around modifying the solvers with different searches and tuning their parameters. 

Thanks for the suggestions. :)

**Comment by johannahaffner** [comment]
Alright! Happy solving :)
